app/routes.py

```diff
@@ -1,2 +1,220 @@
 from flask import Blueprint
+from app import db
+from app.models.customer import Customer
+from app.models.video import Video
+from app.models.rental import Rental
+from flask import request, Blueprint, make_response, jsonify
+from datetime import datetime, timedelta
+import os
+from dotenv import load_dotenv
 
+load_dotenv()
+
+####################### Blueprints ########################
+customers_bp = Blueprint("customers", __name__, url_prefix="/customers")
+videos_bp = Blueprint("videos", __name__, url_prefix="/videos")
+rentals_bp = Blueprint("rentals", __name__, url_prefix="/rentals")
+welcome_page = Blueprint('welcome_page', __name__)
+
+#################### Routes for Welcome page ####################
+@welcome_page.route('/')
+def index():
+    return {
+        "name": "Weishan Yang",
+        "message": "Hi instructors! Welcome to the rainbow video store!"
+    }
+
+#################### Routes for Customers ####################
+@customers_bp.route("", methods=["GET"], strict_slashes=False)
+def get_customer():
+    customers = Customer.query.all()
+    customer_response = [customer.to_json() for customer in customers]
+    return jsonify(customer_response), 200
+
+@customers_bp.route("/<customer_id>", methods=["GET"], strict_slashes=False)  
+def get_single_customer(customer_id):
+    customer = Customer.query.get(customer_id)
+    if customer:
+        return customer.to_json(), 200
+    else:
+        return {"details": "Customer not found"}, 404
+
+@customers_bp.route("/<customer_id>/rentals", methods=["GET"], strict_slashes=False)
+def get_customer_rentals(customer_id):
+    customer = Customer.query.get(customer_id)
+    if not customer:
+        return {"details": "Customer not found"}, 404
+    videos = customer.videos_rent
+    response = []
+    for video in videos:
+        related_rental = Rental.query.filter_by(customer_id=customer_id, video_id=video.video_id).first() 
+        detailed_info = {
+            "release_date": video.release_date,
+            "title": video.title,
+            "due_date": related_rental.due_date
+        }
+        response.append(detailed_info)
+    return jsonify(response), 200
+
+@customers_bp.route("/<customer_id>", methods=["PUT"], strict_slashes=False) 
+def update_customer(customer_id):
+    customer = Customer.query.get(customer_id)
+    if customer:
+        form_data = request.get_json()
+        if all(key in form_data for key in ("name", "postal_code", "phone")):
+            customer.name = form_data["name"]
+            customer.postal_code = form_data["postal_code"]
+            customer.phone = form_data["phone"]
+            db.session.commit()
+            return customer.to_json(), 200
+        else:
+            return {"details": "Invalid data"}, 400
```

Comment:
I suggest you having some validation functions.   They can be used to dry up your code.